src/v/datalake/values_protobuf.cc

```diff
@@ -0,0 +1,333 @@
+              std::get<int32_t>(std::move(field.value())));
+        }
+        co_return iceberg::int_value(
+          field_descriptor.default_value_enum()->number());
+    case pb::FieldDescriptor::TYPE_BOOL:
+        co_return convert<bool, iceberg::boolean_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_bool);
+    case pb::FieldDescriptor::TYPE_STRING:
+        co_return convert<iobuf, iceberg::string_value>(
+          std::move(field),
+          field_descriptor,
+          &pb::FieldDescriptor::default_value_string);
+    case pb::FieldDescriptor::TYPE_GROUP:
+        co_return type_conversion_error(field_descriptor);
+    case pb::FieldDescriptor::TYPE_MESSAGE:
+        if (!field.has_value()) {
+            co_return co_await message_to_value(
+              std::make_unique<parsed::message>(),
+              *field_descriptor.message_type(),
+              stack);
+        }
+        co_return co_await message_to_value(
+          std::get<std::unique_ptr<parsed::message>>(std::move(field.value())),
+          *field_descriptor.message_type(),
+          stack);
+    case pb::FieldDescriptor::TYPE_BYTES:
+        if (!field.has_value()) {
+            co_return iceberg::binary_value{};
+        }
+        co_return iceberg::binary_value(
+          std::get<iobuf>(std::move(field.value())));
+    }
+}
+
+ss::future<value_outcome> message_to_value(
+  std::unique_ptr<parsed::message> message,
+  const pb::Descriptor& descriptor,
+  proto_descriptors_stack& stack) {
+    if (is_recursive_type(descriptor, stack)) {
+        co_return value_conversion_exception(fmt::format(
+          "Recursive message types are not supported. Descriptor: {}",
+          descriptor.DebugString()));
+    }
+    if (stack.size() >= max_recursion_depth) {
+        co_return value_conversion_exception(fmt::format(
+          "Reached maximum recursion depth. Descriptor: {}",
+          descriptor.DebugString()));
+    }
+    stack.push_back(&descriptor);
+    auto ret = std::make_unique<iceberg::struct_value>();
+    ret->fields.reserve(descriptor.field_count());
+
+    /**
+     * The conversion is driven by walking through the message descriptor as
+     * some field with default values are skipped in parsed result.
+     */
+    for (int i = 0; i < descriptor.field_count(); ++i) {
+        auto field_descriptor = descriptor.field(i);
+        auto it = message->fields.find(field_descriptor->number());
+        auto field = it == message->fields.end()
+                       ? std::nullopt
+                       : std::make_optional<parsed::message::field>(
+                           std::move(it->second));
+
+        auto result = co_await message_field_to_value(
+          std::move(field), *field_descriptor, stack);
+
+        if (result.has_error()) {
+            co_return result.error();
+        }
+        ret->fields.push_back(std::move(result.value()));
```

Comment:
@andrwng what is the semantics of a struct with a list of values? How do we translate the list entry to the struct field? I'm worried the order here could not be the order we expect elsewhere when we pair up the schema. AFAIK this ordering is going to be the order in which the fields are written in the source, which is user defined and arbitrary. If the schema needs to match this order, that's a little implicit and we need some comments to that effect.

Reply:
We walk the schema, not the message, so the ordering should be deterministic. Michal's tests should also fail if that changes. The current schema order is not by field number, but by source code order in the .proto file. 

It's just *very* implicit behavior, and it's not a fun invariant to maintain in all translation layers here :shrug: